Remove use of raw pointers in cachedResourceLoader class, IconLoader, ApplicationManifestLoader.

arunsundarapple opened the pull request:
#### ad9f1876e28f6497dfe297d81d3f396efef6b7dc
<pre>
Remove use of raw pointers in cachedResourceLoader class, IconLoader, ApplicationManifestLoader.
<a href="https://bugs.webkit.org/show_bug.cgi?id=254437.">https://bugs.webkit.org/show_bug.cgi?id=254437.</a>
rdar://107196321.

Reviewed by Chris Dumez.

Changed m_documentLoader ivar to CheckedPtr/CheckedRef instead of raw pointer..

* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::requestResource):
(WebCore::CachedResourceLoader::clearDocumentLoader):
(WebCore::CachedResourceLoader::loadDone):
* Source/WebCore/loader/cache/CachedResourceLoader.h:
(WebCore::CachedResourceLoader::clearDocumentLoader): Deleted.

Canonical link: <a href="https://commits.webkit.org/262238@main">https://commits.webkit.org/262238@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/54416fa4f960400d1eea92457f2ee103b60b62f8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/952 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/979 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1016 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1401 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/846 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/941 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1031 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1064 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1054 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/961 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/900 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/895 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1311 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/946 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/892 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/887 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/870 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/919 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/1928 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/910 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/862 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/858 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/901 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/241 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/924 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/ApplicationManifestLoader.h:
```diff
@@ -54,7 +54,7 @@ WTF_MAKE_NONCOPYABLE(ApplicationManifestLoader); WTF_MAKE_FAST_ALLOCATED;
     void notifyFinished(CachedResource&, const NetworkLoadMetrics&);
 
-    DocumentLoader& m_documentLoader;
+    Ref<DocumentLoader> m_documentLoader;
```
This looks like this will introduce a memory leak.
DocumentLoader owns this object via a unique_ptr and now this object keeps a strong ref to its owner.

This means that once DocumentLoader has initialized DocumentLoader::m_applicationManifestLoader, it will never be deletable.

cdumez reviewed (changes requested):
I didn't review the rest of the patch.

cdumez commented on Source/WebCore/loader/ApplicationManifestLoader.h:
```diff
@@ -54,7 +54,7 @@ WTF_MAKE_NONCOPYABLE(ApplicationManifestLoader); WTF_MAKE_FAST_ALLOCATED;
     void notifyFinished(CachedResource&, const NetworkLoadMetrics&);
 
-    DocumentLoader& m_documentLoader;
+    WeakPtr<DocumentLoader> m_documentLoader;
```
WeakPtr is not a great candidate here since m_documentLoader is our owner and cannot be null. CheckedRef<> would be a better candidate. (didn't look at the rest of the patch).

cdumez replied:
WeakPtr is not a great candidate here since m_documentLoader is our owner and cannot be null. CheckedRef<> would be a better candidate. (didn't look at the rest of the patch).

cdumez commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -1019,7 +1019,7 @@ ResourceErrorOr<CachedResourceHandle<CachedResource>> CachedResourceLoader::requ
 #if ENABLE(CONTENT_EXTENSIONS)
         const auto& resourceRequest = request.resourceRequest();
-        auto results = page.userContentProvider().processContentRuleListsForLoad(page, resourceRequest.url(), ContentExtensions::toResourceType(type, request.resourceRequest().requester()), *m_documentLoader);
+        auto results = page.userContentProvider().processContentRuleListsForLoad(page, resourceRequest.url(), ContentExtensions::toResourceType(type, request.resourceRequest().requester()), *m_documentLoader.get());
```
This change shouldn't be needed.

cdumez commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -1568,9 +1568,11 @@ CachePolicy CachedResourceLoader::cachePolicy(CachedResource::Type type, const U
 void CachedResourceLoader::loadDone(LoadCompletionType type, bool shouldPerformPostLoadActions)
 {
-    RefPtr<DocumentLoader> protectDocumentLoader(m_documentLoader);
+    RefPtr<DocumentLoader> protectDocumentLoader(m_documentLoader.get());
```
Since you're modifying this code `<DocumentLoader>` can be omitted, although the variable name doesn't match our usual style. This should be:
```
RefPtr protectedDocumentLoader { m_documentLoader.get() };
```

cdumez reviewed (approved):
LGTM with a few minor changes.

arunsundarapple committed:
Remove use of raw pointers in cachedResourceLoader class, IconLoader, ApplicationManifestLoader.
https://bugs.webkit.org/show_bug.cgi?id=254437.
rdar://107196321.

Reviewed by Chris Dumez.

Changed m_documentLoader ivar to CheckedPtr/CheckedRef instead of raw pointer..

* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::requestResource):
(WebCore::CachedResourceLoader::clearDocumentLoader):
(WebCore::CachedResourceLoader::loadDone):
* Source/WebCore/loader/cache/CachedResourceLoader.h:
(WebCore::CachedResourceLoader::clearDocumentLoader): Deleted.

Canonical link: https://commits.webkit.org/262238@main
